### ISSUE TITLE
Screen readers read "blank" or are silent when changing options in the dropdown

### ISSUE THREAD

**Issue by okonomiyaki** [select2/select2#3735]
I tested the basic example on the examples page with JAWS 16 and NVDA 2015.2 screen readers, and in both cases, the options are not read when expanded and navigating through the options with the up and down arrow keys. JAWS is silent when changing options, and NVDA reads "blank".

Since we already use select2 in production and need to make this accessible, I'd be willing to help test ARIA related modifications.

**Comment by kevin-brown** [comment]
Accessibility is important to us, but unfortunately it's something that we find difficult to test. During the creation of 4.0.0, we closely tested Select2 with GNOME's Orca to make sure it emulated the native select box as closely as possible.

If you could possibly give us some step-by-step instructions on how to test this issue with NVDA, that would be immensely helpful.

**Comment by okonomiyaki** [comment]
I'm fairly new to NVDA myself, but here are some instructions for testing the issue with NVDA:

1. Start NVDA
2. Load examples page (https://select2.github.io/examples.html) in web browser; I tested in Firefox and Chrome. It starts reading the page (type ctrl to tell it to stop).
3. Use tab to navigate to the basic select2 example.
4. alt-down-arrow to expand list. Expected result: list is expanded.
5. up or down arrow to navigate through options. Expected result: options are spoken. Actual result: 'blank' is spoken. 

Compare with the regular select box.

NVDA probably reads 'blank' because it thinks the editable text is focused. Other examples of accessible comboboxes I've seen have used the [aria-owns](http://www.w3.org/TR/wai-aria/states_and_properties#aria-owns) and [aria-activedescendant](http://www.w3.org/TR/wai-aria/states_and_properties#aria-activedescendant) attribute on the editable input which let the screen reader know what has focus.

Here are a couple of examples, albeit with different feature sets than select2:
http://oaa-accessibility.org/examplep/combobox3/
http://jquense.github.io/react-widgets/docs/#/combobox

**Comment by afercia** [comment]
Maybe the OOA example that uses `aria-autocomplete=“inline”` would be more interesting for your needs:
http://oaa-accessibility.org/examplep/combobox2/

**Comment by afercia** [comment]
Please also notice both examples use a container with `role="application"` and the options list uses `role="listbox"` and `role="option"` for the items, that's also what Mr. Zehe suggests to do for autocomplete widgets, see: https://www.marcozehe.de/2014/03/11/easy-aria-tip-7-use-listbox-and-option-roles-when-constructing-autocomplete-lists/

**Comment by okonomiyaki** [comment]
Hi Andrea, I thought about mentioning that when giving the examples, but I didn't have evidence that role="tree" and role="treeitem" could not be made to work as well as role="listbox" and role="option". The treeitem elements appear in MSAA as outline items, not menu items, so the issues that Marco Zehe raises might not apply. 

That said, I'm not certain that a tree structure is necessary to represent a list with groups, unless select2 supports hierarchical lists. This [MDN post](https://developer.mozilla.org/en-US/docs/Web/Accessibility/AT-APIs/Differences/MSAA#Positional_Descriptions_are_Supported) about how Gecko, Firefox's rendering engine, overrides descriptions for outline items suggest that the outline item may result in too much information, but I haven't tested it.

**Comment by afercia** [comment]
Hi okonomiyaki :) This is what a treeview is intended to be:
http://www.w3.org/TR/wai-aria-practices/#TreeView
So it's really about items that can be expanded or collapsed. Working example:
http://accessibleculture.org/research-files/aria-tree-views/treeviewD.html
Tested with Firefox + NVDA and perfectly works, it also announces the "level".
![It works]
(https://cldup.com/jS1f_-rs3U.png)

As I see it, a treeview it's not appropriate to replicate a native select which is basically a list of options with optional groups. All the implementations I've seen use listbox and option for the list of results. I definitely agree the aria-owns and aria-activedescendant properties should be set on the focused field. Heading up to #3744

**Comment by binary-koan** [comment]
Hi, I'm also working on making this accessible - the main change I've made so far is adding aria-activedescendant to the search input and that seems to improve things a lot. I'll try and submit a pull request soon once we've done some more testing, although it probably won't be a comprehensive solution ...

BTW, I tend to test with ChromeVox (at least initially) - as a sighted user it's easiest to have a screen reader which only affects the browser, and in my experience ChromeVox's feature set is much more similar to JAWS/NVDA than Orca's is.

**Comment by gadley** [comment]
I'm still able to reproduce this issue. I'm using version 47.0.2526.80 of chrome and 2015.4 of nvda and version 4.0.1 of select2. The screen reader is still reading blank when changing options in the dropdown. Any ideas from those than have been able to get this to work would be greatly appreciated.

**Comment by ShawnXu** [comment]
Same on JAWS 17. Our current work around is to use an "live" region to hold current text of current highlighted title - when hled use javascript to update the regions text.

**Comment by skerrvy** [comment]
I am able to reproduce this issue on Chrome  48.0.2564.109 and 2015.4 of NVDA and 4.0.1. of select2, however firefox 43.0.4 correctly reads out the items to me using the same versions of NVDA and select2.

**Comment by afercia** [comment]
Getting different results here, can still reproduce on Firefox, see screenshot below (Firefox 44, NVDA 2015.4). Worth noting testing with Chrome + NVDA is a bit pointless, since it's not the best browser/screen reader combination.

![Firefox 44 + NVDA]
(https://cldup.com/lW4_T0efGD.png)

**Comment by afercia** [comment]
@kevin-brown hello. I've recently spent some time investigating on this and I think there's probably more than one reason why the options are not announced by screen readers. I _think_ this would require to proceed step by step, with various PRs to evaluate and hopefully merge.
I'd greatly appreciate any recommendation and any thoughts :)

I'm also trying to draft a sort of broader plan to make Select2 accessible as much as possible. Would greatly appreciate to discuss it with you and anyone who's interested :)

**Comment by afercia** [comment]
Just submitted a PR as a first attempt to improve things :) 

There are best practices and examples to follow to build a "combobox widget", and also the suggested W3C authoring practices, see:
**ARIA authoring practices: combobox**
https://www.w3.org/TR/wai-aria-practices/#combobox
There are two design pattern examples, the second one fits our case. Please consider these are just suggested design patterns, so the "button" mentioned in the examples is not really required. The main things are:
- The first element for the combobox has a role of combobox and behaves like an input text field
- The next element has a listbox role and represents the drop down list
- Each item in the listbox is an option
- Additionally: https://www.w3.org/TR/wai-aria-practices/#Listbox Each selected list item should have aria-selected="true"

I'd suggest to have a look at this post by Mr. James Craig (one of the Editors of the ARIA spec among other things):
**Using ARIA 1.0 and the WebKit Accessibility Node Inspector (Jun 10, 2014)**
https://webkit.org/blog/3302/aria-and-accessibility-inspector/
See Example 3: Combo Box with a “Status” Live Region. Feel free to skip the "live region" part but do watch the video! Important thing to notice: the aria-selected attribute usage.

Also, two examples to follow:
- James Craig demo: https://webkit.org/blog-files/aria1.0/combobox_with_live_region_status.html
- the Twitter search :)

ARIA roles and properties are still implemented inconsistently across browsers and Assistive Technologies so I'd recommend to focus on testing with the main browsers/screen readers combinations, the ones known to work best are:
- Firefox and NVDA
- Safari and VoiceOver
- Internet Explorer 11 and JAWS

This first PR should make the two basic "Single" and "Multiple" examples work with
- Firefox and NVDA: it seems to me everything works fine, even the "optgroups" are announced correctly
- Internet Explorer 11 and JAWS: works but doesn't announces the "optgroups" and there's an issue to solve with initial focus on the "Multiple" select, I'd say this should go in a separate issue

About Safari and VoiceOver, they need `aria-selected` on the "highlighted" option so this would require some additional change, going to open a separate issue for this.

As far as I see, there are a couple more issues to solve, related to missing IDs and a big Safari issue with the nested options, going to open separate issues and reference this ticket.

Any thoughts and feedback are very welcome! Also, testing would be nice :)

TL;DR a working draft with some research and a possible plan:
https://github.com/wpaccessibility/select2-a11y/issues/1

**Comment by ScriptShot** [comment]
Hallo guys, 
i'm very interessted in this progress. Our web application needs to be accsessible, so it would be great if this great component will be accessible shortly. 
Does somebody know how long it will take? Additionaly i want to say: Until now you guys have made a great job! :)

**Comment by ashokramadoss** [comment]
Hellow Guys,

 i am using jaws16 and read as silent when give down arrow key in keyboard for dropdown box.

 its not happing for all the time.

Any solution for this issue

**Comment by Hosch250** [comment]
I have this same problem, as well as a couple others. I have a fix for some of them, but this appears to be caused by the focus being set to the input box even while the highlight is changed in the list. If I filter the results, it reads the filter to me every time. Any thoughts?

**Comment by azizsagi** [comment]
I am having same problem. Options are not readable by jaws. Please help.

**Comment by micker** [comment]
some review on this ?

**Comment by Hosch250** [comment]
Knock knock. Who's there? Your conscience. Wondering if there's a status update on this.

**Comment by ScriptShot** [comment]
i think some people and companies who are using your nice piece would be happy if this problem is not ignored. Accessibility is very important for some people. So please stay tuned.

**Comment by ScriptShot** [comment]
Knock knock, again and again...

**Comment by ecumike** [comment]
We are getting this accessibility error:  "The text input element is not valid for the combobox role."

**Comment by ecumike** [comment]
![image](https://user-images.githubusercontent.com/7328459/60284643-08d54d80-98da-11e9-9847-b24a15e30ea7.png)

**Comment by vlad-saling** [comment]
For reference (and/or if you want to reproduce report above), the tool I and @ecumike use to scan for accessibility issues is this one https://chrome.google.com/webstore/detail/ai-enabled-accessibility/gglkcegocebeligemjccigfaloccnkeo

**Comment by StudioMaX** [comment]
Bump

**Comment by StudioMaX** [comment]
Bump

**Comment by StudioMaX** [comment]
Bump

**Comment by kevin-brown** [comment]
Stale should never come around and mark this ticket again. 🎉

**Comment by Hosch250** [comment]
Thank you. I'd try to create a fix, but I'm not overly familiar with JS, and would rather avoid it if possible.

**Comment by hongwei1** [comment]
iOS VoiceOver screen reader --> A user is able to expand and collapse the dropdown now, however, after expanding and swiping, the focus does not move to the first list item as would be the expected behavior. Instead, the focus moves past these items and into the next element. Only once the user has swiped through the remainder of the page does the focus then jump back into the dropdown list items.

**Comment by c-goetz** [comment]
Maybe this helps somebody running into this issue:
After reading the comment with `aria-activedescendant` I checked if this is set on the input field. It was not. As far as I can see this depends on `_resultId` being set on the item. This was undefined for our usecase. It seems that this should be set in `src/js/select2/data/select.js#_normalizeItem`. This fails because `item._resultId` is undefined instead of null.

So I went for the workaround to just set `_resultId` to the same value as `id` before passing the items to select2.

This works on Windows/Chrome/NVDA and Linux/Firefox/Orca.

**Comment by kevin-brown** [comment]
> In our case an undefined `container` is the cause.

Do you happen to know why `container` was undefined? I've heard of this happening in other cases and I wonder if it's a contributor to this issue still occurring in newer versions of Select2.

**Comment by jamesuk189** [comment]
I had a similar case to @c-goetz where setting `_resultId` seemed to resolve this and enabled the screen reader to begin reading the results.

I can replicate the behaviour in the [AJAX example](https://select2.org/data-sources/ajax) but it doesn't seem to occur in the [search example](https://select2.org/searching).

**Comment by jfolcini** [comment]
Unfortunately I can confirm this bug is still ongoing.

**Comment by kevin-brown** [comment]
The current issue with the `_resultId` not being set appears to be limited to `data` and `ajax` option usage because those both dynamically generate options which are not being normalized properly. Still need to dig deeper to figure out why `container` is not being defined in time, since it should be happening after they are bound to the Select2 container.

**Comment by renoth** [comment]
This problem persists with Wicketstuff select2 latest Version (using select2 4.1.0-rc.0)
The problem is a bit different: NVDA reads the search input when selecting another option with arrow keys. If no search is entered, it reads "blank"

**Comment by fsteffi** [comment]
hi has this issue been resolved

**Comment by fsteffi** [comment]
hi @okonomiyaki has this issue been resolved

**Comment by okonomiyaki** [comment]
> hi @okonomiyaki has this issue been resolved

@fsteffi  I'm no longer working on accessibility issues, so unfortunately I'm unable to say whether this is resolved.

**Comment by msftedad** [comment]
Hi Team, Any update on this issue?
